spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/MetaStore.scala

```diff
@@ -0,0 +1,573 @@
+          }
+        }
+    }
+
+    // Add source tables from Cassandra tables
+    val conn = new CassandraConnector(sqlContext.getCassandraConnConf(clusterName))
+    if (keyspace.nonEmpty) {
+      val ksDef = Schema.fromCassandra(conn).keyspaceByName.get(keyspace.get)
+      names ++= ksDef.map(_.tableByName.keySet).getOrElse(Set.empty).map((name => (name, false)))
+    }
+     names.toList
+  }
+
+  override def getAllDatabases(cluster: Option[String] = None): Seq[String] = {
+    val databaseNamesFromMetastore = getAllDatabasesFromMetastore(cluster).toSet
+    val clusterName = cluster.getOrElse(sqlContext.getCluster)
+    // Add source tables from Cassandra tables
+    val conn = new CassandraConnector(sqlContext.getCassandraConnConf(clusterName))
+    val keyspacesFromCassandra = Schema.fromCassandra(conn).keyspaceByName.keySet
+    (databaseNamesFromMetastore ++ keyspacesFromCassandra -- SystemKeyspaces).toSeq
+  }
+
+  def getAllDatabasesFromMetastore(cluster: Option[String] = None): Seq[String] = {
+    val clusterName = cluster.getOrElse(sqlContext.getCluster)
+    val selectQuery =
+      s"""
+      |SELECT keyspace_name
+      |From ${getMetaStoreTableFullName}
+      |WHERE cluster_name = '$clusterName'
+    """.stripMargin.replaceAll("\n", " ")
+    val names = ListBuffer[String]()
+    // Add source tables from metastore
+    metaStoreConn.withSessionDo {
+      session =>
+        val result = session.execute(selectQuery).iterator()
+        while (result.hasNext) {
+          val row: Row = result.next()
+          val keyspaceName = row.getString(0)
+          if (keyspaceName != TempDatabaseOrTableToBeDeletedName) {
+            names += keyspaceName
+          }
+        }
+    }
+    names.toList
+  }
+
+  override def getAllClusters(): Seq[String] = {
+    (getAllClustersFromMetastore ++ Seq(sqlContext.getCluster)).distinct
+  }
+
+  def getAllClustersFromMetastore: Seq[String] = {
+    val selectQuery =
+      s"""
+      |SELECT cluster_name
+      |From ${getMetaStoreTableFullName}
+    """.stripMargin.replaceAll("\n", " ")
+    val names = ListBuffer[String]()
+    // Add source tables from metastore
+    metaStoreConn.withSessionDo {
+      session =>
+        val result = session.execute(selectQuery).iterator()
+        while (result.hasNext) {
+          val row: Row = result.next()
+          val clusterName = row.getString(0)
+          names += clusterName
+        }
+    }
+
+    names.toList
+  }
+
+  /** Store a tale with the creation meta data */
+  override def storeTable(
+      tableIdent: TableIdent,
+      source: String,
+      schema: Option[StructType],
+      options: Map[String, String]): Unit = {
+    import collection.JavaConversions._
+
+    val cluster = tableIdent.cluster.getOrElse(sqlContext.getCluster)
+    if (schema.nonEmpty) {
+      metaStoreConn.withSessionDo {
+        session =>
+          val preparedStatement = session.prepare(InsertIntoMetaStoreQuery)
+          session.execute(
```

Comment:
This pattern is repeated in many places of code, which means it is a nice candidate for being abstracted out to a separate method. The call site should look like:
```scala
execute(
  InsertIntoMetaStoreQuery, 
  cluster, 
  tableIdent.keyspace, 
  tableIdent.table, 
  source,
  ...
```